### PR TITLE
Get contract name in ToJson

### PR DESCRIPTION
**Pull request by superboyiii**
Related: https://github.com/neo-project/neo-modules/pull/378
Partly support: https://github.com/neo-project/neo-modules/issues/355
Enable get name in ToJson.

### CODE REVIEW

**Comment by superboyiii** [comment]
> Related: [neo-project/neo-modules#378](https://github.com/neo-project/neo-modules/pull/378)
> Partly support: [neo-project/neo-modules#355](https://github.com/neo-project/neo-modules/issues/355)
> Enable get native name in ToJson.

**Comment by roman-khimov** [comment]
We might want to extend manifests to include some name for all contracts, btw. Neo 2 had that metadata and it allowed contract explorers to show something more meaningful than just hashes.

**Comment by superboyiii** [comment]
> We might want to extend manifests to include some name for all contracts, btw. Neo 2 had that metadata and it allowed contract explorers to show something more meaningful than just hashes.

Yes, might add it later.

File: src/neo/Ledger/ContractState.cs [hunk 1]
```diff
@@ -74,6 +75,14 @@ public JObject ToJson()
         {
             JObject json = new JObject();
             json["id"] = Id;
+            if (Id < 0)
+            {
+                foreach (var native in NativeContract.Contracts)
+                {
+                    if (Id == native.Id)
+                        json["name"] = native.Name;
```
Comment (Tommo-L) [inline src/neo/Ledger/ContractState.cs:83]:
```suggestion
                   {
                        json["name"] = native.Name;
                        break;
                    }
```

**Comment by superboyiii** [comment]
Done

**Review by Tommo-L — APPROVED** [review]

**Comment by superboyiii** [comment]
@shargon Shall we merge?

**Comment by shargon** [comment]
Please take a look, I don't like to make difference between native and not native.

**Comment by erikzhang** [comment]
May I ask why you need to get the name of the contract in json? Isn't this public information? They can be hard-coded into the SDK.

**Comment by roman-khimov** [comment]
I think any contract should have a name.

**Comment by erikzhang** [comment]
> I think any contract should have a name.

I also think so. But how should they display their names? Some contracts use the `name()` method, and some use the manifest. If all contracts are required to have a name, then a unified way must be used to display the names.

**Comment by superboyiii** [comment]
> May I ask why you need to get the name of the contract in json? Isn't this public information? They can be hard-coded into the SDK.

As https://github.com/neo-project/neo-modules/issues/355 said, although hash is necessary, name/ID is more convenient to remember as one of the contract infos. I think we have duty to make it more friendly for development. And as @roman-khimov said, maybe we could make name and contact as required when deploying as neo2 did (neo2 get name in contract state from what they input when deploying).

**Comment by superboyiii** [comment]
> > May I ask why you need to get the name of the contract in json? Isn't this public information? They can be hard-coded into the SDK.
> 
> As [neo-project/neo-modules#355](https://github.com/neo-project/neo-modules/issues/355) said, although hash is necessary, name/ID is more convenient to remember as one of the contract infos. I think we have duty to make it more friendly for development. And as @roman-khimov said, maybe we could make name and contact as required when deploying as neo2 did (neo2 get name in contract state from what they input when deploying) or we could make a check on name from manifest to ensure it contains name and return error msg when it's null.

**Comment by superboyiii** [comment]
> Please take a look, I don't like to make difference between native and not native.

Agree

**Comment by roman-khimov** [comment]
> Some contracts use the name() method, and some use the manifest

Manifests are probably more appropriate for this, both native and non-native contracts could have a field there for name.

**Comment by erikzhang** [comment]
> Manifests are probably more appropriate for this, both native and non-native contracts could have a field there for name.

Then we should modify NEP-5 and other standards, and remove the `name()` method.

**Comment by shargon** [comment]
I think that it will be easier with name method, I can change the code to execute `name` if we are agree

**Comment by roman-khimov** [comment]
`name()` method requires spawning a VM, running some code and then dealing with the output, to me it's a bit harder than getting data from a field in the manifest. At the same time, `name()` is something available to other contracts unlike manifest, but that can be solved with some `System.Contract.GetManifest` syscall probably.

**Comment by Tommo-L** [comment]
> although hash is necessary, name/ID is more convenient to remember as one of the contract infos.

Wouldn't it be more convenient to use NNS.
 
 For me, we'd better keep `name` method, like NEP-5/11, we already have `symbol`, `decimals` methods. If just move `name` method to manifest file, iIt doesn't look consistent.

**Comment by Tommo-L** [comment]
Now, we can move on

**Comment by shargon** [comment]
> Now, we can move on
yes, we can take it from the manifest

**Comment by superboyiii** [comment]
@roman-khimov @shargon @erikzhang Please review again.

File: src/neo/Ledger/ContractState.cs [hunk 2]
```diff
@@ -71,6 +72,30 @@ public JObject ToJson()
...
+                {
+                    json["name"] = Manifest.Name;
+                }
+
+                json["isnative"] = false;
+            }
+
             json["hash"] = ScriptHash.ToString();
             json["script"] = Convert.ToBase64String(Script);
             json["manifest"] = Manifest.ToJson();
```
Comment (erikzhang) [inline src/neo/Ledger/ContractState.cs:101]: The name is included in the `manifest`, and I think we don't need to add it to the top level of the json.

Comment (superboyiii) [reply src/neo/Ledger/ContractState.cs:101]: > The name is included in the `manifest`, and I think we don't need to add it to the top level of the json.

Agree. I'll close this.

**Pull request closed without merging**